### PR TITLE
Add date picker to load traces for a specific date in the future

### LINKED ISSUE #43: Add a date picker for traces date selection
# Story
As a Traces app user I'd like to be able to select the date from which I want to see the traces starting. Currently it is only possible to see the dates starting from `today`.

## Goal
The date picker would allow a user to select the date from which the traces are loaded.

# Implementation
In order to implement this, each page should have a date picker that serves as the starting point from which traces are shown.

## Current things that can be reused
1. In the `TracesCollectorService` the functions to get the traces already take DateTime `from` and `to` as parameters to determine the date interval that should be loaded.
2. As for each page a different function needs to be called for instance, to load for a reservation you need to use the `LoadTracesForReservationAsync` function. So you should follow the same implementation path as for instance the `LoadTracesAsync` function where it is just an `abstract` function in the `TracesBaseViewmodel` and then the implementation is in each specific ViewModel (hint: your function should be a public one, in order to use bind and be called from the UI)
3. From your new function when the user changes the date with the picker, you can call directly the `LoadTracesAsync` function and have it reload all the traces with the new interval.

## Remarks
1. On the first load, the selected date should be today.
1. If possible, use the `Blazorise` `DateEdit` element [Blazorise DateEdit Component](https://blazorise.com/docs/components/date/) this component has a `DateChanged` event, that would prove helpful to call the function in the ViewModel.
1. When the user changes the date, the current list of traces should be cleared and loaded from the new date onwards.
2. When a user selects a new date, only the selected date, and the following day should be loaded in the beginning. The load more traces button should allow the user to keep moving forward from that date onwards, but we don't want to clutter the page with too much.
4. User should only be able to select dates from today in the future. (No time traveling to the past)
3. When a user selects a date that is in the future (not today), the overdue list of traces should be cleared and not show at all. Subsequently when the user goes back to the date of today then overdue traces can be reloaded.

### PR DESCRIPTION
**Pull request by haydar-metin**
closes #43

it also adds icons

image
image

### CODE REVIEW

File: src/Traces.Web/ViewModels/TracesBaseViewModel.cs [hunk 1]
```diff
@@ -201,6 +202,8 @@ public void HideEditTraceModal()
 
         public abstract Task LoadNextDaysAsync();
 
+        public abstract Task LoadFromDateAsync(DateTime date);
```
Comment (randyrod) [inline src/Traces.Web/ViewModels/TracesBaseViewModel.cs]: I noticed that the implementation of this function in each viewmodel is the same, so there is no need for it to be `abstract`.

You can just have the implementation of this function in this class only, and avoid duplicate code, or did I miss something?

File: src/Traces.Web/ViewModels/TracesPropertyViewModel.cs [hunk 2]
```diff
@@ -95,6 +95,27 @@ public override async Task LoadNextDaysAsync()
             }
         }
 
+        public override async Task LoadFromDateAsync(DateTime from)
```
Comment (randyrod) [inline src/Traces.Web/ViewModels/TracesPropertyViewModel.cs]: Same as in `TracesAccountViewModel`

File: src/Traces.Web/ViewModels/TracesReservationViewModel.cs [hunk 3]
```diff
@@ -99,6 +99,27 @@ public override async Task LoadNextDaysAsync()
             }
         }
 
+        public override async Task LoadFromDateAsync(DateTime from)
```
Comment (randyrod) [inline src/Traces.Web/ViewModels/TracesReservationViewModel.cs]: Same as in `TracesAccountViewModel`

File: src/Traces.Web/ViewModels/TracesAccountViewModel.cs [hunk 4]
```diff
@@ -60,6 +66,27 @@ public override async Task LoadNextDaysAsync()
             }
         }
 
+        public override async Task LoadFromDateAsync(DateTime from)
```
Comment (randyrod) [inline src/Traces.Web/ViewModels/TracesAccountViewModel.cs]: This code is repeated and can be moved to `TracesBaseViewModel`

Refer to comment in `TracesBaseViewModel` for further information.

File: src/Traces.Web/Shared/DatePicker.razor [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    private DateTime DatePickerDate { get; set; } = DateTime.Today;
+
+    private bool IsPast => DatePickerDate < DateTime.Today;
+    private bool IsPreviousDayDisabled => DatePickerDate.AddDays(-1) < DateTime.Today;
+    private bool IsPreviousWeekDisabled => DatePickerDate.AddDays(-7) < DateTime.Today;
+    
+    private async Task OnDateChange(DateTime? date)
+    {
+        DateTime newDate = date.GetValueOrDefault(DateTime.Today);
```
Comment (randyrod) [inline src/Traces.Web/Shared/DatePicker.razor]:
```suggestion
        var newDate = date.GetValueOrDefault(DateTime.Today);
```

File: src/Traces.Web/Shared/DatePicker.razor [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+        if (newDate >= DateTime.Today)
+        {
+            await OnChange.InvokeAsync(date.GetValueOrDefault(DateTime.Today));
```
Comment (randyrod) [inline src/Traces.Web/Shared/DatePicker.razor]:
```suggestion
            await OnChange.InvokeAsync(newDate);
```

File: src/Traces.Web/Shared/DatePicker.razor [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        else
+        {
+            Date = newDate;
+        }
+    }
+    
+    private async Task SetToday()
```
Comment (randyrod) [inline src/Traces.Web/Shared/DatePicker.razor]:
```suggestion
    private async Task SetTodayButtonOnClick()
```
Just a suggestion not a requirement

File: src/Traces.Web/Shared/DatePicker.razor [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        await OnDateChange(DateTime.Today);
+    }
+
+    private async Task NextDay()
```
Comment (randyrod) [inline src/Traces.Web/Shared/DatePicker.razor]:
```suggestion
    private async Task NextDayButtonOnClick()
```

Just a suggestion, not a requirement

File: src/Traces.Web/Shared/DatePicker.razor [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        await OnDateChange(DatePickerDate.AddDays(1));
+    }
+
+    private async Task NextWeek()
```
Comment (randyrod) [inline src/Traces.Web/Shared/DatePicker.razor]:
```suggestion
    private async Task NextWeekButtonOnClick()
```

File: src/Traces.Web/Shared/DatePicker.razor [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        await OnDateChange(DatePickerDate.AddDays(7));
+    }
+
+    private async Task PreviousDay()
```
Comment (randyrod) [inline src/Traces.Web/Shared/DatePicker.razor]:
```suggestion
    private async Task PreviousDayButtonOnClick()
```

File: src/Traces.Web/Shared/DatePicker.razor [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        await OnDateChange(DatePickerDate.AddDays(-1));
+    }
+
+    private async Task PreviousWeek()
```
Comment (randyrod) [inline src/Traces.Web/Shared/DatePicker.razor]:
```suggestion
    private async Task PreviousWeekButtonOnClick()
```

File: src/Traces.Web/ViewModels/TracesBaseViewModel.cs (hunk 1 above)
Comment (haydar-metin) [reply src/Traces.Web/ViewModels/TracesBaseViewModel.cs]: i followed your issue, i wanted to discuss it with you, i didnt know you are already back :+1: i will change it

File: src/Traces.Web/Shared/DatePicker.razor [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+    .date-picker {
+        
+    }
+
+</style>
+
+<div class="py-3 date-picker">
+    <div class="row no-gutters">
+        <Button Clicked="@SetToday">
```
Comment (randyrod) [inline src/Traces.Web/Shared/DatePicker.razor]: The `Today` doesn't look like an actionable element. At least on first glance. I would give it some love, maybe you can make it with the primary color, but outlined to see how it looks. See the blazorise doc for more info on what I mean https://bootstrapdemo.blazorise.com/tests/buttons

File: src/Traces.Web/Shared/DatePicker.razor (hunk 5 above)
Comment (haydar-metin) [reply src/Traces.Web/Shared/DatePicker.razor]: cool like in java :)

File: src/Traces.Web/Shared/DatePicker.razor (hunk 6 above)
Comment (haydar-metin) [reply src/Traces.Web/Shared/DatePicker.razor]: :man_facepalming:

**Pull request merged**